### PR TITLE
holes

### PR DESCRIPTION
**Pull request by soenkehahn**
This is in preparation for #147.

### CODE REVIEW

**Comment by soenkehahn** [comment]
@hallettj: Good catch, I've changed the PR. Could you try again, please?

File: src/recorder/hole_recorder.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            } => {
+                if !checker.protocol.steps.is_empty() {
+                    checker.handle_execve_enter(pid, registers, executable, arguments)
+                } else {
+                    match checker.result {
+                        CheckerResult::Failure(_) => {
+                            checker.handle_execve_enter(pid, registers, executable, arguments)
+                        }
+                        CheckerResult::Pass => {
+                            *self = HoleRecorder::Recorder {
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply src/recorder/hole_recorder.rs]: Yeah, I don't know any other language where this could show up... :)

File: src/utils/mod.rs [hunk 2]
```diff
@@ -22,3 +23,23 @@ pub fn parse_hashbang(program: &Path) -> Option<String> {
...
+        fn next(&mut self) -> Option<Self::Item> {
+            let result = self.0.next();
+            match result {
+                None => None,
+                Some(current) => Some((current, self.0.peek().is_some())),
+            }
+        }
+    }
+
+    Iter(into_iter.into_iter().peekable())
```
Comment (matthandlersux) [inline src/utils/mod.rs]: `into_iter.into_iter()` reads funny, is there a better name for the argument to `with_has_more`?

**Comment by matthandlersux** [comment]
i wasn't fully done with the review but wanted to get the comments in there before heading out ☝️

File: README.md [hunk 3]
```diff
@@ -165,6 +165,33 @@ protocols:
...
+the specified holes:
+
+``` yaml
+protocols:
+  - arguments: foo
+    protocol:
+      - git add .
+      - _
+```
+
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply README.md]: I added a section to the readme. Is that better?

File: src/lib.rs [hunk 4]
```diff
@@ -128,129 +126,23 @@ pub fn run_check_protocols(context: &Context, script: &Path) -> R<ExitCode> {
...
-            &Context::new_mock(),
-            &None,
-            &script.path(),
-            Protocol::empty(),
-            &[],
-        )?;
-        assert!(!testfile.path().exists(), "touch was executed");
-        Ok(())
-    }
+    let (protocols_file, expected_protocols) = Protocols::load(script)?;
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply src/lib.rs]: Done.

File: src/recorder/mod.rs [hunk 5]
```diff
@@ -1,28 +1,30 @@
...
     command: Option<Command>,
 }
 
-impl Default for Recorder {
-    fn default() -> Recorder {
+impl Recorder {
+    pub fn new(protocol: Protocol) -> Recorder {
         Recorder {
-            protocol: Protocol::new(vec![]),
+            protocol,
```
[the remark this answers is not in the document]
Comment (soenkehahn) [reply src/recorder/mod.rs]: Changed it back. The new constructor is called `new_with_protocol` now.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (matthandlersux) [reply README.md]: 👍

File: src/utils/mod.rs (hunk 2 above)
Comment (soenkehahn) [reply src/utils/mod.rs]: Inside `with_has_more` the only thing we know about the argument is that it implements `IntoIterator`. So I thought `into_iter` was the best name. What would be better? `input`? `iterable`? Or just `iterator`?

**Pull request merged**